modules/post/linux/manage/adduser.rb

```diff
@@ -0,0 +1,274 @@
+        host = framework.db.workspace.hosts.where(address: session.session_host).first
+        if host.os_name == 'linux' && host.os_flavor
+          host.os_flavor
+        else
+          host.os_name
+        end
+      else
+        get_sysinfo[:distro]
+      end
+
+    binary_case = case binary
+                  when /useradd$/
+                    'useradd'
+                  when /adduser$/
+                    'adduser'
+                  else
+                    binary
+                  end
+    case binary_case
+    when 'useradd'
+      print_status("Running on #{os_platform}")
+      print_status('Useradd exists. Using that')
+      case os_platform
+      when /debian|ubuntu|fedora|centos|oracle|redhat|arch|suse|gentoo/i
+        homedirc = datastore['HOME'].empty? ? '--no-create-home' : "--home-dir #{datastore['HOME']}"
+
+        # Since command can add on groups, checking over groups
+        groupadd = ''
+        if datastore['MissingGroups'] == 'CREATE'
+          groupadd = command_exists?('groupadd') ? 'groupadd' : 'addgroup'
+
+          groups_missing.each do |group|
+            d_cmd_exec("#{groupadd} #{group}")
+            print_good("Added #{group} group")
+          end
+        end
+        groupsc = groups.empty? ? '' : "--groups #{groups.join(',')}"
+
+        # Finally run it
+        d_cmd_exec("#{binary} --password \'#{passwd}\' #{homedirc} #{groupsc} --shell #{datastore['SHELL']} --no-log-init #{datastore['USERNAME']}".gsub(/ {2,}/, ' '))
+        groups_handled = true
+      else
+        vprint_status('Unsure what platform were on. Using useradd in most basic/common settings')
+
+        # Finally run it
+        d_cmd_exec("#{binary} #{datastore['USERNAME']} | echo")
+        d_cmd_exec("echo \'#{datastore['USERNAME']}:#{passwd}\'|chpasswd -e")
+      end
+    when 'adduser'
+      print_status("Running on #{os_platform}")
+      print_status('Adduser exists. Using that')
+      case os_platform
+      when /debian|ubuntu/i
+        print_bad('Adduser cannot add groups to the new user automatically. Going to have to do it at a later step')
+        homedirc = datastore['HOME'].empty? ? '--no-create-home' : "--home #{datastore['HOME']}"
+
+        d_cmd_exec("#{binary} --disabled-password #{homedirc} --shell #{datastore['SHELL']} #{datastore['USERNAME']} | echo")
+        d_cmd_exec("echo \'#{datastore['USERNAME']}:#{passwd}\'|chpasswd -e")
+      when /fedora|centos|oracle|redhat/i
+        homedirc = datastore['HOME'].empty? ? '--no-create-home' : "--home-dir #{datastore['HOME']}"
+
+        # Since command can add on groups, checking over groups
+        groupadd = ''
+        if datastore['MissingGroups'] == 'CREATE'
+          groupadd = command_exists?('groupadd') ? 'groupadd' : 'addgroup'
+
+          groups_missing.each do |group|
+            d_cmd_exec("#{groupadd} #{group}")
+            print_good("Added #{group} group")
+          end
+        end
```

Comment:
We could have the user attached to each group independent, but part of why I dont is because each OS's user binary is unique. They all have different settings and different arguments. Its a miracle for the few consistent arguments they have which get thrown out  the window on another system. If I know that a systems useradd can add groups, I'd rather deal with that than have another spiderweb to handle attaching groups.

That being said, I could add another advanced option that requires a trigger of attaching the users at the end and just skips the groups portion of the command, but its more compact and simpler when dealing with each one individually since each ones its own beast

Reply:
It makes sense. I was just worried about future code maintenance due to the complexity. I think if you can refactor this as I suggested this will help a lot. I would avoid adding one more advanced option to keep it simple.

Reply:
So, I think I found a good compromise. I noticed that the section that I have to create groups was present in every conditional branch in some way, so I just moved the section actually creating the nonexistant groups towards the top with the rest of the group molding, which meant all the other repeats are gone now. Linking is still separate, but its a single line in each conditional rather than the code block it was before. The changes are here 0c005a50c90ca6f78a039d0aedd52c7c9cc343a8

Reply:
Actually in addition, I did move the whole groups section into its own function